add java i386 arch

gabemontero opened the pull request:
@bparees PTAL

*** NOTE *** 

I wanted to see what it would take to use `alternatives` to manage the java version as the `1001` user, since `alternatives` is a more complete switch (handles everything, not just the `java` binary).

Here are the updates for centos and jenkins v1 and v2.

Based on how the review goes, I'll circle back and start looking into updating the rhel image.

bparees commented on 1/Dockerfile:
```diff
@@ -49,6 +49,35 @@ RUN /usr/local/bin/plugins.sh /opt/openshift/base-plugins.txt && \
     chown -R 1001:0 /opt/openshift && \
+    chown -R 1001:0 /etc/alternatives && \
+    chown -R 1001:0 /var/lib/alternatives && \
+    chown -R 1001:0 /usr/lib/jvm && \
```
what are we doing at assemble time (that's the only thing that runs under uid 1001) that needs ownership of these directories?

bparees commented on 1/Dockerfile:
```diff
@@ -49,6 +49,35 @@ RUN /usr/local/bin/plugins.sh /opt/openshift/base-plugins.txt && \
+    chown -R 1001:0 /usr/lib/jvm && \
+    chmod 775 /usr/bin && \
+    chmod 775 /usr/lib/jvm-exports && \
+    chmod 775 /usr/share/man/man1 && \
```
what are we doing in these dirs that the existing permissions were problematic?

bparees replied:
(and when are we doing it?  assemble or run time?)

bparees commented on 1/Dockerfile:
```diff
@@ -49,6 +49,35 @@ RUN /usr/local/bin/plugins.sh /opt/openshift/base-plugins.txt && \
+    unlink /usr/share/man/man1/rmiregistry.1.gz && \
+    unlink /usr/share/man/man1/servertool.1.gz && \
+    unlink /usr/share/man/man1/tnameserv.1.gz && \
+    unlink /usr/share/man/man1/unpack200.1.gz && \
```
do we need to unlink all these files?  the man pages in particular i'd guess we don't care about.  for the bin directory, does "alternatives" actually setup content in the bin directory?  Could we simply make sure the actual JVM paths appear in $PATH before /usr/bin?

gabemontero replied:
I incrementally added each `unlink` until I was able to get the `alternatives` command to work.  It manipulates all of the above file system content.

bparees commented:
I think most of my questions boil down to "exactly what does "alternatives -set java" do?

bparees commented on 1/test/run:
```diff
@@ -103,7 +103,7 @@ function test_get_job() {
 function run_tests() {
   local name=$1 ; shift
-  CONTAINER_ARGS="-u 10000:0"
+  CONTAINER_ARGS="-u 1001:0"
```
(from your IRC query) we're setting it to 10000 to ensure it can run as a random uid, not the uid we built the image as.

if it only passes when you set the container uid to 1001, then it's going to fail when run on openshift.

gabemontero commented:
OK, so the `alternatives` command (do a `man alternatives` on say your fedora laptop, or see https://access.redhat.com/documentation/en-US/JBoss_Enterprise_Application_Platform/6/html/Installation_Guide/Install_OpenJDK_on_Red_Hat_Enterprise_Linux.html) allows you to manage all the binaries of a given "application" and switch between versions when multiple versions can exist on a system.
Java is such an "application".

Essentially every binary I am unlinking in the Dockerfile is a binary under the Java umbrella that is managed by `alternatives`.

Per the doc noted above, it is assumed you'll be root or using sudo, so all this was the attempt to work around that.

Ultimately, `/usr/bin/java` is linked to `/etc/alternatives/java`, which is then linked to the specific version of java you want (i.e. i386 vs. x86_64, or specifhttps://access.redhat.com/documentation/en-US/JBoss_Enterprise_Application_Platform/6/html/Installation_Guide/Install_OpenJDK_on_Red_Hat_Enterprise_Linux.htmlic versions of either of those two arch's).  The `alternatives` command will replace the links for all the files listed under the Java "family" to the version you specify.

Simply changing the permissions was not sufficient though.  `alternatives` does some `ln` type operations under the covers.  Evidently when running as root, you can existing links.  But you can't `ln` over an existing file when running as non-root, independent of the permission of the file.  So the `ln` portion of `alternatives` would fail.  Unlinking in the Dockerfile prior to running `alternatives` in the run script was the workaround.

As to the random ID, to work around that, we would ditch the `chown` calls and replace them with `chmod 777` to allow the random IDs to manipulate these files.

So the question:  is this extent of permission change the lesser or greater evil, as compared to using JAVA_HOME or fully resolved path to `java` in our various scripts, but having all the other java-related binaries point to the other arch.  My concern with the latter is how assured can we be the Jenkins or some plugin will pull in won't have a dependency on one of those other binaries.

Or we simply just switch to 32 bit and don't provide a dynamic switch.

gabemontero commented:
OK I've reverted the test run scripts to use the random ID and updated the Dockerfiles to leverage chmod to open things up further to users beyond root or 1001.  `make test` passes locally for me.

bparees commented:
> So the question: is this extent of permission change the lesser or greater evil, as compared to using JAVA_HOME or fully resolved path to java in our various scripts, but having all the other java-related binaries point to the other arch. My concern with the latter is how assured can we be the Jenkins or some plugin will pull in won't have a dependency on one of those other binaries.

yeah i certainly like using `alternatives` better than relying on java_home (it's never been clear to me to what extent that actually controls the classpath and other aspects of the application runtime)

bparees commented on 1/Dockerfile:
```diff
@@ -48,6 +48,35 @@ RUN /usr/local/bin/plugins.sh /opt/openshift/base-plugins.txt && \
+    chmod -R 777 /usr/lib/jvm && \
+    chmod 777 /usr/bin && \
+    chmod 777 /usr/lib/jvm-exports && \
+    chmod 777 /usr/share/man/man1 && \
```
chgrp'ing to 0 should also work (as long as the dir has group rwx permissions). 

I don't know that it's actually any more secure at the end of the day, though it feels slightly more secure.

gabemontero replied:
Gotcha ... let me see about getting a stable PR test run with the current approach.  If we get to that point, I'll see about switching up.

gabemontero replied:
turns out the group only has r-x permissions:

```
drwxr-xr-x.  2 root root 4096 Jan 25 19:01 man1
```

Leaving things as they are wrt chmod, unless I'm misinterpreting your previous comment.

bparees replied:
yeah i meant that we could chmod them to 775 (since they are already in the root group, there is no need to chgrp them)

gabemontero replied:
OK ... switching to 775 did in fact work for me when running the tests.  Must have misinterpreted something earlier.

Pushed the updates.

bparees commented on 1/contrib/s2i/run:
```diff
@@ -9,6 +9,29 @@
 
+#display the available i386 and x86_64 java's to confirm their presence and version
+alternatives --display java | grep family | grep i386
+alternatives --display java | grep family | grep x86_64
```
is this just for debugging purposes?  doesn't seem like output we'd likely want in the final image.

gabemontero replied:
Originally debug, but then I left in cause I thought it might be useful when debugging the odd ball customer issue that could be a JDK issue ... channeling my prior existence as an app server developer where knowing the precise JDK version was important :-)

bparees replied:
yeah, I can certainly appreciate the "let's dump all the debug info we want by default" attitude, but i think we should be able to get that info if we really need it relatively easily in this case.

bparees commented on 1/contrib/s2i/run:
```diff
@@ -9,6 +9,29 @@
+fi
+
+#confirm what version of java is being used
+ls -la /etc/alternatives/java
```
same question here (debugging?)

gabemontero replied:
Basic / same response as above  .... could see getting rid of the ones above and just keeping this one for that debug note on which JDK the customer went with

bparees replied:
how about dump both, but expand the echo statements above?  "Using JDK $FOO based on OPENSHIFT_JENKINS_JVM_ARCH value of $OPENSHIFT_JENKINS_JVM_ARCH"

bparees commented:
we should add a test that brings up the image using the 32 bit jvm.

gabemontero commented:
The 32 bit jvm comes up, but an initial foray into a 32 bit jvm test brought up a potential issue.  The password related stuff failed, and I now noticed this in the jenkins log:

```
ERROR: ld.so: object 'libnss_wrapper.so' from LD_PRELOAD cannot be preloaded: ignored.
```

Looks like we'll need to see if a 32 bit version of the nss stuff is out there.

gabemontero commented:
OK I've pushed an update that 
- adds a 32 bit test
- tweaks the debug along the lines discussed
- installs (albeit via rpmfind.net which may not be a long term solution) 32 bit nss for centos

`make test` passes for me locally

I'll circle back to the `chgrp` comment, then look at 32-bit for `slave-base` and the headless java stuff.

gabemontero commented:
On Fri, Jan 27, 2017 at 4:37 PM Ben Parees <notifications@github.com> wrote:

> *@bparees* commented on this pull request.
> ------------------------------
>
> In 1/Dockerfile <https://github.com/openshift/jenkins/pull/238>:
>
> > @@ -48,6 +48,35 @@ RUN /usr/local/bin/plugins.sh /opt/openshift/base-plugins.txt && \
>      touch /opt/openshift/plugins/subversion.jpi.pinned && \
>      touch /opt/openshift/plugins/ssh-credentials.jpi.pinned && \
>      touch /opt/openshift/plugins/script-security.jpi.pinned && \
> +    chmod -R 777 /etc/alternatives && \
> +    chmod -R 777 /var/lib/alternatives && \
> +    chmod -R 777 /usr/lib/jvm && \
> +    chmod 777 /usr/bin && \
> +    chmod 777 /usr/lib/jvm-exports && \
> +    chmod 777 /usr/share/man/man1 && \
>
> yeah i meant that we could chmod them to 775 (since they are already in
> the root group, there is no need to chgrp them)
>
ah ok i originally tried 775 and it didn't work with the random 10000 uid

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/jenkins/pull/238>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadErkCYXcRNflwkGh7OG0X2K_J1zdks5rWmOXgaJpZM4LuAsj>
> .
>

bparees commented:
> ah ok i originally tried 775 and it didn't work with the random 10000 uid

strange.  the container user is always a member of the root group.

gabemontero commented:
Pretty sure that was the case, but I did iterate over the solution quickly.  Maybe I misinterpreted something.  I'll retry when I get the chance to make sure.

Also, I just pushed the slave-base changes for 32 bit centos.

bparees commented on 1/Dockerfile:
```diff
@@ -28,9 +28,12 @@ EXPOSE 8080 50000
+    INSTALL_PKGS="rsync gettext git tar zip unzip java-1.8.0-openjdk java-1.8.0-openjdk.i686 java-1.8.0-openjdk-devel java-1.8.0-openjdk-devel.i686 jenkins-1.651.3-1.1 nss_wrapper wget" && \
     yum -y --setopt=tsflags=nodocs install $INSTALL_PKGS && \
     rpm -V $INSTALL_PKGS && \
+    wget ftp://rpmfind.net/linux/fedora/linux/development/rawhide/Everything/x86_64/os/Packages/n/nss_wrapper-1.1.3-1.fc25.i686.rpm && \
```
yeah i definitely want to work out a better solution with @tdawson (but it's fine for testing/proving it can work).

bparees commented on 1/test/run:
```diff
@@ -102,12 +102,16 @@ function test_get_job() {
+  if [ $# -gt 1 ];
+  then
+      CONTAINER_ARGS+=" -e OPENSHIFT_JENKINS_JVM_ARCH=i386 "
+  fi
```
we handle this more generically in other test code:
https://github.com/sclorg/mongodb-container/blob/master/3.2/test/run#L163-L167
https://github.com/sclorg/mongodb-container/blob/master/3.2/test/run#L291-L299

gabemontero replied:
Didn't rework to exactly like the mongo run's (would have lead to a much broader structural change) but I reworked it to something I think is similar.  Changes pushed.

bparees commented:
looks ok to me, but i still want us to pursue the centos nss wrapper rpm before merging.

gabemontero commented:
Absolutely - I've also started rhel changes.  Or do we only want to merge
centos initially?

On Mon, Jan 30, 2017 at 4:53 PM Ben Parees <notifications@github.com> wrote:

> looks ok to me, but i still want us to pursue the centos nss wrapper rpm
> before merging.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/jenkins/pull/238#issuecomment-276203585>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadGkm4Kw-rMmHbWlUxy_KG8_jlQ3lks5rXlvcgaJpZM4LuAsj>
> .
>

bparees commented:
starting rhel makes sense.

gabemontero commented:
OK, I've pushed the rhel changes after some verification on a local AWS rhel instance I have.

@tdawson - if you recall the 32 bit nss_wrapper rpms for centos7 and rhel7 discussion you @bparees and I had on IRC a few days ago, do you have a preference on how we "officially" engage with you to get those RPMs available on a repo for our jenkins build to access on your plate?  Or is it too soon to "officially" engage :-) ?

thanks

gabemontero commented:
Hmm ... the use of `alternatives` failed in the PR test for rhel even though it passed for me in my local aws instance.

Going to temporarily add some debug and see what is up.

gabemontero commented:
Also taking out SKIP_SQUASH based on our recent email thread.

gabemontero commented:
Ah .. did not pull all the changes from my AWS instance over to my local repo (since I can't get vagrant-openshift sync for the jenkins repo to work yet :-( )

Pushing fix in a sec.

gabemontero commented:
OK, now pushed update to dynamically set to 32bit if mem is less that 4GB and the env var is not explicitly set to use 64 bit.  PTAL @bparees

bparees commented on 1/contrib/s2i/run:
```diff
@@ -9,11 +9,36 @@
+
+# set the java version used based on OPENSHIFT_JENKINS_JVM_ARCH
+if [ -z $OPENSHIFT_JENKINS_JVM_ARCH  ]; then
+    if [ "${CONTAINER_MEMORY_IN_BYTES}" -gt "${MEMORY_LIMIT_FOR_32BIT}" ]; then
```
unfortunately it's more complicated than this.

What the check really needs to be based on is whether the computed heap size is greater than what a 32 bit jvm allows.  Depending on CONTAINER_HEAP_PERCENT, the computed heap could be significantly less than the container memory.

also what is the actual max value a 32 bit jvm on linux allows the heap to be set to?  is it 2**32-1?  i would think it might be a bit lower than that since some process memory needs to be reserved for native usage.

bparees replied:
btw i'm perfectly ok w/ not doing this magic.

gabemontero commented:
Ok I can see about adding more sophistication.  Though I wonder if it
really needs to be that precise.  If the user is manipulating the heap with
the other props and hitting edge cases he can just add the prop to
explicitly turn on 32 bit.

But yeah if it becomes a rabbit hole we can just punt.

On Tue, Jan 31, 2017 at 5:41 PM Ben Parees <notifications@github.com> wrote:

> *@bparees* commented on this pull request.
> ------------------------------
>
> In 1/contrib/s2i/run <https://github.com/openshift/jenkins/pull/238>:
>
> > @@ -9,11 +9,36 @@
>  source /usr/local/bin/jenkins-common.sh
>  source /usr/local/bin/kube-slave-common.sh
>
> +#get the fully qualified paths to both 32 and 64 bit java
> +JVMPath32bit=`alternatives --display java | grep family | grep i386 | awk '{print $1}'`
> +JVMPath64bit=`alternatives --display java | grep family | grep x86_64 | awk '{print $1}'`
> +
> +#if machine has 4GB or less, going to force use of 32bit to save space unless user
> +#specifically want to force 64bit
> +CONTAINER_MEMORY_IN_BYTES=`cat /sys/fs/cgroup/memory/memory.limit_in_bytes`
> +MEMORY_LIMIT_FOR_32BIT=$((2**32-1))
> +
> +# set the java version used based on OPENSHIFT_JENKINS_JVM_ARCH
> +if [ -z $OPENSHIFT_JENKINS_JVM_ARCH  ]; then
> +    if [ "${CONTAINER_MEMORY_IN_BYTES}" -gt "${MEMORY_LIMIT_FOR_32BIT}" ]; then
>
> btw i'm perfectly ok w/ not doing this magic.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/jenkins/pull/238>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadE_ATCeYBqfes_gnssP2MdumD3SOks5rX7iygaJpZM4LuAsj>
> .
>

gabemontero commented:
Pushed the changes for the 32 vs. 64 bit auto tuning per our discussion in scrum this AM.

gabemontero commented:
OK pushed the changes to remove nss_wrapper as a dependency per discussion in scrum this AM.

Assuming this holds we won't need our previous ask of @tdawson regarding getting 32 bit nss_wrapper available on our repos during image build.

bparees commented on hack/build.sh:
```diff
@@ -27,9 +27,6 @@ function docker_build_with_version {
   echo "LABEL io.openshift.builder-version=\"${git_version}\"" >> "${dockerfile}.version"
   docker build -t ${IMAGE_NAME} -f "${dockerfile}.version" .
-  if [[ "${SKIP_SQUASH}" != "1" ]]; then
-    squash "${dockerfile}.version"
```
you can remove the actual squash function too.

gabemontero replied:
update pushed

bparees commented on 1/Dockerfile:
```diff
@@ -48,6 +48,36 @@ RUN /usr/local/bin/plugins.sh /opt/openshift/base-plugins.txt && \
     touch /opt/openshift/plugins/subversion.jpi.pinned && \
     touch /opt/openshift/plugins/ssh-credentials.jpi.pinned && \
     touch /opt/openshift/plugins/script-security.jpi.pinned && \
+    chmod 775 /etc/passwd && \
```
i'm having second thoughts about this since it means anyone who can run a job on jenkins can update the passwd file which in turn means they can have free run of the jenkins container, including changing the jenkins config to give themselves more access.  (for example I think they could add/update an existing user with a different password and then su to that user)

it would be better if we can "close the door behind us" after we update the file on startup but i don't think that's going to be possible.

going to have to think about this one some more.

gabemontero replied:
@bparees and I had a confab with Clayton ... that coupled with some experimentation leads us to believe we are OK, but are still going to sit and ponder for a bit since this PR is post 3.5.

tdawson commented:
Thanks for removing the nss_wrapper dependency.  Every scenario I could think of was rather messy.

bparees commented on 1/test/run:
```diff
@@ -192,6 +191,16 @@ function run_s2i_tests() {
 
 # Normal tests
+echo "Normal test: 64 bit"
+CONTAINER_ARGS="-u 10000:0"
```
should set the JVM explicitly to 64bit, because if we (somehow) run this test on a machine w/ low memory it's going to go through the 32 bit path.

gabemontero replied:
good catch ... part of next push

bparees commented on 2/test/run:
```diff
@@ -192,6 +191,16 @@ function run_s2i_tests() {
 
 # Normal tests
+echo "Normal test: 64 bit"
+CONTAINER_ARGS="-u 10000:0"
```
same thing here

bparees commented:
@gabemontero one last test related nit and then lgtm.  We'll hold this until the 3.5 jenkins image goes out the door though.

gabemontero commented:
Updates to test pushed.

btw, more scm-api plugin churn seen during the v2 test:

```
SEVERE: Failed Loading plugin GitHub Branch Source Plugin v2.0.3 (github-branch-source)
java.io.IOException: GitHub Branch Source Plugin v2.0.3 failed to load.
 - SCM API Plugin v2.0.3 is older than required. To fix, install v2.0.4 or later.
```

I'll create a separate PR, then rebase this PR after that PR merges.

gabemontero commented:
ok test passing again after plugin update rebase

gabemontero commented:
@bparees  bump ... are we close enough to 3.5 GA to update the centos image and have @tdawson cook up a rhel7 image on brew for QA to use?  If not there yet, are we close enough that a precise target date is known?

thanks

bparees commented:
@gabemontero probably the end of this sprint, which is when we hit codefreeze for 3.5.

gabemontero commented:
On Tue, Feb 21, 2017 at 11:41 AM Ben Parees <notifications@github.com>
wrote:

> @gabemontero <https://github.com/gabemontero> probably the end of this
> sprint, which is when we hit codefreeze for 3.5.
>
makes sense ok thx for update

> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/jenkins/pull/238#issuecomment-281400384>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadObWxiGZcyVU1VOFtpzZX2K9NNTHks5rexOigaJpZM4LuAsj>
> .
>
